pipeline/data/importers/mono/hplt.py

```diff
@@ -0,0 +1,219 @@
+        )
+        self.document_count = CountingStep(
+            dataset_path,
+            "How many documents were visited. This can help represent data diversity.",
+        )
+        self.final_lines = CountingStep(
+            dataset_path,
+            "How many lines were actually written. Smaller lines will be combined together.",
+        )
+
+    def count_shards_visited(self):
+        self.shards.filtered -= 1
+        self.shards.kept += 1
+
+
+def language_has_hplt_support(language: str) -> bool:
+    return location_exists(
+        f"https://data.hplt-project.org/one/monotext/cleaned/{language}_map.txt"
+    )
+
+
+def load_shard_urls(language: str) -> list[str]:
+    """
+    Download the list of shards, e.g.
+
+    https://data.hplt-project.org/one/monotext/cleaned/en/en_100.jsonl.zst
+    https://data.hplt-project.org/one/monotext/cleaned/en/en_101.jsonl.zst
+    https://data.hplt-project.org/one/monotext/cleaned/en/en_102.jsonl.zst
+    https://data.hplt-project.org/one/monotext/cleaned/en/en_103.jsonl.zst
+    https://data.hplt-project.org/one/monotext/cleaned/en/en_104.jsonl.zst
+    ...
+    https://data.hplt-project.org/one/monotext/cleaned/en/en_110.jsonl.zst
+    """
+
+    url = f"https://data.hplt-project.org/one/monotext/cleaned/{language}_map.txt"
+    logger.info(f"Downloading shard list: {url}")
+
+    with read_lines(url) as lines:
+        shard_urls = []
+        for line in lines:
+            shard_urls.append(line.strip())
+    random.Random(url).shuffle(shard_urls)
+
+    logger.info(f"Available shards for {language}:")
+    for lines in shard_urls:
+        logger.info(f" - {lines}")
+    return shard_urls
+
+
+def download_hplt(
+    language: str,
+    min_fluency_threshold: float,
+    max_lines: int,
+    max_words_in_sentence,
+    file_destination: Path,
+):
+    """
+    Downloads and filters the HPLT dataset.
+    https://hplt-project.org/datasets/v1.2
+
+    In the English data with a fluency score of 0.9, about 5% of the total sentences visited
+    are fluent enough to keep. That means to generate a monolingual dataset of 100 million
+    lines 2 billion lines need to be visited.
+
+    Parameters:
+     - language: The BCP 47 language code to filter the documents.
+     - min_fluency_threshold: The minimum score a sentence must have to be included in the final dataset.
+     - max_lines: The maximum number of lines to include in the final dataset.
+     - max_words_in_sentence: The maximum number of words allowed in each sentence.
+     - file_destination: The destination path where the final dataset will be written.
+    """
+
+    with ExitStack() as stack:
+        stats = FilteringStatistics(file_destination)
+
+        outfile = stack.enter_context(write_lines(file_destination))
+
+        shard_urls = load_shard_urls(language)
+        stats.shards.filtered = len(shard_urls)
+        document_stream = stack.enter_context(
```

Comment:
Could you explain how this works? Does it just read the shuffled shards one by one?